File: ch01/python_inheritance/ws.py
import time
import smbus
import flask

class Sensor:
  def __init__(self):
    self.measurement = -999

  def get_read_as_string(self):
    return('{} {}'.format(
      self.measurement, self.units))

class Si7021Temp(Sensor):
  def __init__(self, name, bus_id):
    super().__init__()
    self.name = name
    self.units = 'Celsius'
    self.bus = smbus.SMBus(bus_id)
    
  def measure(self):
    self.bus.write_byte(0x40, 0xF3)
    time.sleep(0.1)
    byte0 = self.bus.read_byte(0x40)
    byte1 = self.bus.read_byte(0x40)
    self.measurement = (
      (byte0 * 256 + byte1) * 
        175.72 / 65536.0) - 46.85
    time.sleep(0.1)

class Si7021Humidity(Sensor):
  def __init__(self, name, bus_id):
    super().__init__()
    self.name = name
    self.units = '%'
    self.bus = smbus.SMBus(bus_id)
    
  def measure(self):
    self.bus.write_byte(0x40, 0xF5)

    time.sleep(0.1)
    byte0 = self.bus.read_byte(0x40)
    byte1 = self.bus.read_byte(0x40)
    self.measurement = (
      (byte0 * 256 + byte1) * 
        125 / 65536.0) - 6
    time.sleep(0.1)


class MPL3115A2Pressure(Sensor):
  def __init__(self, name, bus_id):
    super().__init__()
    self.name = name
    self.units = 'kPa'
    self.bus = smbus.SMBus(bus_id)
  
  def measure(self):
    self.bus.write_byte_data(0x60, 0x26, 0x39)
    time.sleep(0.1)
    data = self.bus.read_i2c_block_data(
      0x60, 0x00, 4)
    self.measurement = (
      (data[1] * 65536) + (data[2] * 256) + 
        (data[3] & 0xF0)) / 64000

sensor_list = []
sensor_list.append(Si7021Temp('temp', 1))
sensor_list.append(Si7021Humidity('humidity', 1))
sensor_list.append(
  MPL3115A2Pressure('air_pressure', 1))

def measuring_loop():
  while True:
    for sensor in sensor_list:
      sensor.measure()
      print('{} measured {:.2f} {}'.format(
        sensor.name, 
        sensor.measurement, 
        sensor.units))
    time.sleep(5)

measuring_loop()

File: ch04/weather-station/ws.js
var express = require('express')
var app = express();
var gpio = require('rpi-gpio');
gpio.setMode(gpio.MODE_BCM);

app.set('view engine', 'pug');
app.use(express.static('static'))
app.get('/dashboard', (request, response) => {
  response.render(
    'dashboard', { sensors: sensorList, fans: fanList });
});

app.get('/gauges', (request, response) => {
  response.render(
    'gauges', { sensors: sensorList, fans: fanList });
});
  
app.listen(3000, function () {
  console.log('Example app listening on port 3000!')
});

const i2c = require("i2c-bus");

class Sensor {
  constructor(name) {
    this.name = name;
    this.measurement = -999;
  }

  toJSON() {
    return {
      name: this.name,
      measurement: this.measurement,
      units: this.units,
      time: this.time
    };
  }
  
  getReadAsString() {
    if (this.hasOwnProperty("units")) { 
      return(this.measurement + ' ' + this.units);
    } else {
      return(String(this.measurement));
    }
  }
}

class Si7021Temp extends Sensor {
  constructor(name, busId) {
    super(name);
    this.units = "deg C";
    this.busId = busId;
    this.i2cBus = i2c.openSync(busId);
  }

  read() {
    this.i2cBus.writeByte(0x40, 0xF3, 0, (err) => {
      if(err) {
        console.log(err);
        this.measurement = -999;
        return;
      }
      setTimeout(() => {
        this.i2cBus.i2cRead(0x40, 3, new Buffer(3),
  (err, bytesRead, data) => {
          if(err) {
            this.measurement = -999;
            return;
          }
          this.measurement = (((((data[0] << 8) | data[1]) * 175.72) / 65536) - 46.85);
          this.time = new Date();
        });
      }, 30);
    });
  }
}

class MPL3115A2Temp extends Sensor {
  constructor(name, busId) {
    super(name);
    this.units = "deg C";
    this.busId = busId;
    this.i2cBus = i2c.openSync(busId);
  }

  read() {
    this.i2cBus.writeByteSync(0x60, 0x26, 0xB9);
    this.i2cBus.writeByteSync(0x60, 0x13, 0x07);
    this.i2cBus.writeByte(0x60, 0x26, 0xB9, (err) => {
      setTimeout(() => {
        let data = new Buffer(6);
        this.i2cBus.i2cReadSync(0x60, 6, data);
        let temp = ((data[4] * 256) + (data[5] & 0xF0)) / 16;
        this.measurement = (temp / 16.0);
        this.time = new Date();
      }, 1000);
    })
  }
}

class Fan {
  constructor(name, pin) {
    this.name = name;
    this.pin =  pin;
    gpio.setup(this.pin, gpio.DIR_OUT);
  }

  set(state) {
    this.state = state;
    gpio.write(this.pin, this.state);
    wsServer.broadcast(JSON.stringify({
      fans: [this]
    }));
  } 
}

const WebSocket = require('ws');
const wsServer = new WebSocket.Server({
  port: 3001,
});

wsServer.broadcast = function(data) {
  for (client of wsServer.clients) {
    client.send(data);
  }
};
 
wsServer.on('connection', function connection(ws) {
  wsServer.broadcast(JSON.stringify({sensors: sensorList}));
  ws.on('message', function incoming(data) {
    //
  });
});


  var sensorList = [];
  var fanList = [];
  var threshold  = 24;
  sensorList.push(new Si7021Temp("Si7021Temp", 1));
  sensorList.push(new MPL3115A2Temp("MPL3115A2Temp", 1));
  fanList.push(new Fan("left", 20));
  fanList.push(new Fan("right", 21));

  setInterval(() => {
    fanList[0].set(sensorList[0].measurement > threshold);
    fanList[1].set(sensorList[1].measurement > threshold);
  }, 1000);

setInterval(() => {
  for (s of sensorList) {
    s.read()  
    wsServer.broadcast(JSON.stringify({
      sensors: [s]
    }));
  } 
}, 1000);

